File: server/src/test/java/de/zalando/zally/dto/ViolationsCounterTest.java
package de.zalando.zally.dto;

import de.zalando.zally.rule.Result;
import de.zalando.zally.rule.api.Severity;
import de.zalando.zally.rule.api.Rule;
import de.zalando.zally.rule.zalando.AvoidTrailingSlashesRule;
import de.zalando.zally.rule.zalando.ZalandoRuleSet;
import org.junit.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import static org.junit.Assert.assertEquals;


public class ViolationsCounterTest {
    private static void assertCounters(Map<Severity, Integer> expectedCounters, List<Result> violations)
        throws AssertionError {

        final ViolationsCounter counter = new ViolationsCounter(violations);
        for (Severity violationType : expectedCounters.keySet()) {
            assertEquals(expectedCounters.get(violationType), counter.getCounter(violationType));
        }
    }

    @Test
    public void returnsZerosWhenViolationListIsEmpty() {
        final List<Result> violations = new ArrayList<>();
        final Map<Severity, Integer> expectedCounters = new HashMap<>();
        expectedCounters.put(Severity.MUST, 0);
        expectedCounters.put(Severity.SHOULD, 0);
        expectedCounters.put(Severity.MAY, 0);
        expectedCounters.put(Severity.HINT, 0);

        assertCounters(expectedCounters, violations);
    }

    @Test
    public void countsMustViolations() {
        final List<Result> violations = new ArrayList<>();
        IntStream.range(0, 5).forEach(
            (i) -> violations.add(generateViolation(Severity.MUST))
        );
        final Map<Severity, Integer> expectedCounters = new HashMap<>();
        expectedCounters.put(Severity.MUST, 5);
        expectedCounters.put(Severity.SHOULD, 0);
        expectedCounters.put(Severity.MAY, 0);
        expectedCounters.put(Severity.HINT, 0);

        assertCounters(expectedCounters, violations);
    }

    @Test
    public void countsShouldViolations() {
        final List<Result> violations = new ArrayList<>();
        IntStream.range(0, 5).forEach(
            (i) -> violations.add(generateViolation(Severity.SHOULD))
        );
        final Map<Severity, Integer> expectedCounters = new HashMap<>();
        expectedCounters.put(Severity.MUST, 0);
        expectedCounters.put(Severity.SHOULD, 5);
        expectedCounters.put(Severity.MAY, 0);
        expectedCounters.put(Severity.HINT, 0);
        assertCounters(expectedCounters, violations);
    }

    @Test
    public void countsMayViolations() {
        final List<Result> violations = new ArrayList<>();
        IntStream.range(0, 5).forEach(
            (i) -> violations.add(generateViolation(Severity.MAY))
        );
        final Map<Severity, Integer> expectedCounters = new HashMap<>();
        expectedCounters.put(Severity.MUST, 0);
        expectedCounters.put(Severity.SHOULD, 0);
        expectedCounters.put(Severity.MAY, 5);
        expectedCounters.put(Severity.HINT, 0);

        assertCounters(expectedCounters, violations);
    }

    @Test
    public void countsHintViolations() {
        final List<Result> violations = new ArrayList<>();
        IntStream.range(0, 5).forEach(
            (i) -> violations.add(generateViolation(Severity.HINT))
        );
        final Map<Severity, Integer> expectedCounters = new HashMap<>();
        expectedCounters.put(Severity.MUST, 0);
        expectedCounters.put(Severity.SHOULD, 0);
        expectedCounters.put(Severity.MAY, 0);
        expectedCounters.put(Severity.HINT, 5);

        assertCounters(expectedCounters, violations);
    }

    private Result generateViolation(Severity violationType) {
        return new Result(new ZalandoRuleSet(),
                AvoidTrailingSlashesRule.class.getAnnotation(Rule.class),
                "Test Description",
                violationType,
                "#/pointer");
    }
}

File: server/src/test/java/de/zalando/zally/dto/ViolationsCounterTest.kt
package de.zalando.zally.dto

import de.zalando.zally.rule.Result
import de.zalando.zally.rule.api.Rule
import de.zalando.zally.rule.api.Severity
import de.zalando.zally.rule.api.Severity.HINT
import de.zalando.zally.rule.api.Severity.MAY
import de.zalando.zally.rule.api.Severity.MUST
import de.zalando.zally.rule.api.Severity.SHOULD
import de.zalando.zally.rule.zalando.AvoidTrailingSlashesRule
import de.zalando.zally.rule.zalando.ZalandoRuleSet
import org.junit.Assert.assertEquals
import org.junit.Test

class ViolationsCounterTest {

    @Test
    fun returnsZerosWhenViolationListIsEmpty() {
        val results = emptyList<Result>()
        val counts = hashMapOf(
            MUST to 0,
            SHOULD to 0,
            MAY to 0,
            HINT to 0
        )

        assertCounters(counts, results)
    }

    @Test
    fun countsMustViolations() {
        val results = listOfResults(5, MUST)
        val counts = hashMapOf(
            MUST to 5,
            SHOULD to 0,
            MAY to 0,
            HINT to 0
        )

        assertCounters(counts, results)
    }

    @Test
    fun countsShouldViolations() {
        val results = listOfResults(4, SHOULD)
        val counts = hashMapOf(
            MUST to 0,
            SHOULD to 4,
            MAY to 0,
            HINT to 0
        )

        assertCounters(counts, results)
    }

    @Test
    fun countsMayViolations() {
        val results = listOfResults(3, MAY)
        val counts = hashMapOf(
            MUST to 0,
            SHOULD to 0,
            MAY to 3,
            HINT to 0
        )

        assertCounters(counts, results)
    }

    @Test
    fun countsHintViolations() {
        val results = listOfResults(2, HINT)
        val counts = hashMapOf(
            MUST to 0,
            SHOULD to 0,
            MAY to 0,
            HINT to 2
        )

        assertCounters(counts, results)
    }

    @Test
    fun countsMixedViolations() {
        val results =
            listOfResults(1, MUST) +
                listOfResults(2, SHOULD) +
                listOfResults(5, HINT)
        val counts = hashMapOf(
            MUST to 1,
            SHOULD to 2,
            MAY to 0,
            HINT to 5
        )

        assertCounters(counts, results)
    }

    private fun listOfResults(count: Int, severity: Severity): List<Result> {
        val ruleSet = ZalandoRuleSet()
        return List(count) {
            Result(
                ruleSet,
                AvoidTrailingSlashesRule::class.java.getAnnotation(Rule::class.java),
                "Test Description",
                severity,
                "/pointer"
            )
        }
    }

    private fun assertCounters(expected: Map<Severity, Int>, results: List<Result>) {
        val counter = ViolationsCounter(results)
        for (violationType in expected.keys) {
            assertEquals(expected[violationType], counter.getCounter(violationType))
        }
    }
}
